File: example/client/memory.js
import { MemuClient } from 'memu-js';
import { readFile } from 'node:fs/promises';
import path from 'node:path';

async function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function normalizeConversationShape(item) {
  // Accept multiple shapes and normalize to either string or array of { role, content }
  if (typeof item === 'string') return item;
  if (Array.isArray(item)) return item; // assume already [{ role, content }, ...]
  if (item && typeof item === 'object') {
    if (typeof item.conversation_text === 'string') return item.conversation_text;
    if (Array.isArray(item.conversation)) return item.conversation;
    if (Array.isArray(item.messages)) return item.messages;
  }
  // Fallback: stringify unknown structures
  return JSON.stringify(item);
}

async function loadConversationsFromJsonFile(filePath) {
  const content = await readFile(filePath, 'utf-8');
  const data = JSON.parse(content);
  const list = Array.isArray(data) ? data : Array.isArray(data?.conversations) ? data.conversations : [];
  return list.map(normalizeConversationShape).filter((v) => v && (typeof v === 'string' || Array.isArray(v)));
}

async function waitForTaskCompletion(client, taskId) {
  // Poll task status until terminal state
  while (true) {
    const status = await client.getTaskStatus(taskId);
    console.log(`Task status: ${status.status}`);
    if (['SUCCESS', 'FAILURE', 'REVOKED'].includes(status.status)) break;
    await sleep(2000);
  }
}

async function main() {
  const client = new MemuClient({
    baseUrl: 'https://api.memu.so',
    apiKey: process.env.MEMU_API_KEY,
  });

  const USER_ID = 'user001';
  const USER_NAME = 'User 001';
  const AGENT_ID = 'assistant001';
  const AGENT_NAME = 'Assistant 001';

  const conversationFile = path.join(path.dirname(new URL(import.meta.url).pathname), 'conversation.json');
  const conversations = await loadConversationsFromJsonFile(conversationFile);

  if (!conversations.length) {
    console.log('No conversations loaded. Exiting.');
    if (typeof client.close === 'function') client.close();
    return;
  }

  console.log(`Processing ${conversations.length} English dialogues...`);

  for (let i = 0; i < conversations.length; i += 1) {
    const conversation = conversations[i];
    console.log(`\nProcessing dialogue ${i + 1}/${conversations.length}`);

    const memoResponse = await client.memorizeConversation(
      conversation, // string or [{ role, content }]
      USER_ID,
      USER_NAME,
      AGENT_ID,
      AGENT_NAME
    );

    console.log(`Saved! Task ID: ${memoResponse.taskId}`);
    await waitForTaskCompletion(client, memoResponse.taskId);
    console.log(`Dialogue ${i + 1} completed successfully!`);
  }

  console.log(`\nAll ${conversations.length} dialogues have been processed and saved to MemU!`);

  // Test recall
  console.log(`\nTesting memory recall...`);
  const memories = await client.retrieveRelatedMemoryItems({
    userId: USER_ID,
    query: 'hiking safety',
    topK: 3,
  });
  console.log(`Found ${memories.totalFound} memories`);
  for (const memoryItem of memories.relatedMemories || []) {
    const content = memoryItem?.memory?.content || '';
    console.log(`Memory: ${content.slice(0, 100)}...`);
  }

  // Retrieve default categories
  console.log('\nRetrieving default categories...');
  const defaultCategories = await client.retrieveDefaultCategories({
    userId: USER_ID,
    agentId: AGENT_ID,
    includeInactive: false,
  });
  console.log(`Found ${defaultCategories.totalCategories} default categories:`);
  for (const category of defaultCategories.categories || []) {
    console.log(`  - ${category.name}: ${category.description}`);
  }

  // Retrieve related clustered categories
  console.log("\nRetrieving related categories for 'outdoor activities'...");
  const relatedCategories = await client.retrieveRelatedClusteredCategories({
    userId: USER_ID,
    categoryQuery: 'outdoor activities',
    topK: 5,
    minSimilarity: 0.3,
  });

  const relatedCount = Array.isArray(relatedCategories.clusteredCategories)
    ? relatedCategories.clusteredCategories.length
    : (relatedCategories.totalCategoriesFound || 0);
  console.log(`Found ${relatedCount} related categories:`);
  for (const category of relatedCategories.clusteredCategories || []) {
    console.log(`  - ${category.name}`);
  }

  if (typeof client.close === 'function') client.close();
}

// Run
main().catch((err) => {
  console.error('Error running memory client:', err);
  process.exitCode = 1;
});

File: example/client/memory.py
import os
import json
import time
from typing import List, Dict

from memu import MemuClient


def load_conversations_from_file(file_path: str) -> List[Dict[str, str]]:
    with open(file_path, 'r', encoding='utf-8') as f:
        conversation = json.load(f)
    return conversation


def wait_for_task_completion(memu_client: MemuClient, task_id: str) -> None:
    """Wait for a memorization task to complete."""
    while True:
        status = memu_client.get_task_status(task_id)
        print(f"Task status: {status.status}")
        if status.status in ['SUCCESS', 'FAILURE', 'REVOKED']:
            break
        time.sleep(2)


def main():
    # Initialize MemU client
    memu_client = MemuClient(
        base_url="https://api.memu.so", 
        api_key=os.getenv("MEMU_API_KEY")
    )

    # Load conversation from JSON file
    conversation_file = os.path.join(os.path.dirname(__file__), "conversation.json")
    conversation_messages = load_conversations_from_file(conversation_file)
    
    if not conversation_messages:
        print("No conversation loaded. Exiting.")
        return


    # Save conversation to MemU
    print("\nProcessing multi-turn conversation")
    memo_response = memu_client.memorize_conversation(
        conversation=conversation_messages,
        user_id="user002", 
        user_name="User 002", 
        agent_id="assistant002", 
        agent_name="Assistant 002"
    )
    print(f"Saved! Task ID: {memo_response.task_id}")

    time.sleep(5)
    # Wait for completion
    wait_for_task_completion(memu_client, memo_response.task_id)
    print("Conversation completed successfully!")

    print("\nConversation has been processed and saved to MemU!")

    memories = memu_client.retrieve_related_memory_items(
        user_id="user001", 
        query="hiking safety", 
        top_k=3
    )
    for memory_item in memories.related_memories:
        print(f"Memory: {memory_item.memory.content[:100]}...")

    # Retrieve default categories
    default_categories = memu_client.retrieve_default_categories(
        user_id="user002",
        agent_id="assistant002",
        include_inactive=False
    )
    print(f"Found {default_categories.total_categories} default categories:")
    for category in default_categories.categories:
        print(f"  - {category.name}: {category.description}")

    # Retrieve related clustered categories
    related_categories = memu_client.retrieve_related_clustered_categories(
        user_id="user002",
        category_query="outdoor activities",
        top_k=5,
        min_similarity=0.3
    )

    print(f"Found {len(related_categories.clustered_categories)} related categories:")
    for category in related_categories.clustered_categories:
        print(f"  - {category.name}")

    # Example: Delete memories
    print("\n🗑️ Deleting memories examples:")
    
    # Delete memories for a specific user and agent
    print("Deleting memories for user002 and assistant002...")
    delete_response = memu_client.delete_memories(
        user_id="user002",
        agent_id="assistant002"
    )
    print(f"✅ Success: {delete_response.success}")
    print(f"   Deleted {delete_response.deleted_count} memories")

    # Example: Delete all memories for a user (without specifying agent)
    print("\nDeleting all memories for user002...")
    delete_all_response = memu_client.delete_memories(user_id="user002")
    print(f"✅ Success: {delete_all_response.success}")
    print(f"   Deleted {delete_all_response.deleted_count} memories")

    memu_client.close()


if __name__ == "__main__":
    main()